File: stream.py

"""
A bit stream class with convenience methods
"""
data = []
pos = 0

def initialize(data_param):
    """
    Initializes the required global variables.
    """
    global data
    data = data_param
    print(id(data))

def GetBit():
    """
    returns the bit in position pos
    """
    global pos
    b = data[pos >> 3]
    s = 7 - (pos & 0x7)
    pos += 1
    return (b >> s) & 1

def GetBitN(l):
    """
    returns l number of bits from position pos
    """
    val = 0
    for _ in range(l):
        val = (val << 1) + GetBit()
    return val

File: decoder-c/stream.cpp
//
// Created by Ayon Debnath on 2024-03-09.
//
#include <iostream>
#include <vector>

int GetBit(std::vector<unsigned char>& data, int& pos) {
    /*
     * returns the bit in position pos from the data variable
     */
    int b = data[pos >> 3];
    int s = 7 - (pos & 0x7);
    pos++;
    return (b >> s) & 1;
}

int GetBitN(int l, std::vector<unsigned char>& data, int& pos) {
    /*
     * returns l number of bits from position pos from the data variable
     */
    int val = 0;
    for (int i = 0; i < l; ++i) {
        val = (val << 1) + GetBit(data, pos);
    }
    return val;
}